src/contracts/permissions/KeyRegistrar.sol

```diff
@@ -0,0 +1,408 @@
+     * @param operator Address of the operator
+     * @param keyData The ECDSA address encoded as bytes (20 bytes)
+     * @param signature EIP-712 signature over the registration message
+     * @dev Validates address format, verifies signature ownership, and ensures global uniqueness
+     */
+    function _registerECDSAKey(
+        OperatorSet memory operatorSet,
+        address operator,
+        bytes calldata keyData,
+        bytes calldata signature
+    ) internal {
+        // Validate ECDSA address format
+        require(keyData.length == 20, InvalidKeyFormat());
+
+        // Decode address from bytes
+        address keyAddress = address(bytes20(keyData));
+        require(keyAddress != address(0), ZeroPubkey());
+
+        // Calculate key hash using the address
+        bytes32 keyHash = _getKeyHashForKeyData(keyData, CurveType.ECDSA);
+
+        // Check global uniqueness
+        require(!globalKeyRegistry[keyHash], KeyAlreadyRegistered());
+
+        // Create EIP-712 compliant message hash
+        bytes32 structHash = keccak256(
+            abi.encode(
+                ECDSA_KEY_REGISTRATION_TYPEHASH, operator, operatorSet.avs, operatorSet.id, keyAddress, block.chainid
+            )
+        );
+        bytes32 messageHash = _hashTypedDataV4(structHash);
+
+        // Verify that the signature was created by the key address
+        address recoveredSigner = ECDSA.recover(messageHash, signature);
+        require(recoveredSigner == keyAddress, InvalidSignature());
+
+        // Store key data using internal helper
+        _storeKeyData(operatorSet, operator, keyData, keyHash);
+    }
+
+    /**
+     * @notice Validates and registers a BN254 public key with proper signature verification
+     * @param operatorSet The operator set to register the key for
+     * @param operator Address of the operator
+     * @param keyData The BN254 public key bytes (G1 and G2 components)
+     * @param signature Signature proving key ownership
+     * @dev Validates keypair, verifies signature using hash-to-G1, and ensures global uniqueness
+     */
+    function _registerBN254Key(
+        OperatorSet memory operatorSet,
+        address operator,
+        bytes calldata keyData,
+        bytes calldata signature
+    ) internal {
+        BN254.G1Point memory g1Point;
+        BN254.G2Point memory g2Point;
+
+        {
+            // Decode BN254 G1 and G2 points from the keyData bytes
+            (uint256 g1X, uint256 g1Y, uint256[2] memory g2X, uint256[2] memory g2Y) =
+                abi.decode(keyData, (uint256, uint256, uint256[2], uint256[2]));
+
+            // Validate G1 point
+            g1Point = BN254.G1Point(g1X, g1Y);
+            require(!(g1X == 0 && g1Y == 0), ZeroPubkey());
+
+            // Construct BN254 G2 point from coordinates
+            g2Point = BN254.G2Point(g2X, g2Y);
+        }
+
+        // Create EIP-712 compliant message hash
+        bytes32 structHash = keccak256(
```

Comment:
need view functions for these for offchain integrations to work

Reply:
+1 Need something similar to `pubkeyRegistrationMessageHash` in `SlashingRegistryCoordinator`.